src/VirtoCommerce.CatalogPersonalizationModule.Data/Migrations/20200914134605_AddTaggedItemUniqueIndex.cs

```diff
@@ -1,11 +1,67 @@
-﻿using Microsoft.EntityFrameworkCore.Migrations;
+using Microsoft.EntityFrameworkCore.Migrations;
 
 namespace VirtoCommerce.CatalogPersonalizationModule.Data.Migrations
 {
     public partial class AddTaggedItemUniqueIndex : Migration
     {
         protected override void Up(MigrationBuilder migrationBuilder)
         {
+            migrationBuilder.Sql(@"
+	            -- 1. Set Tag.TaggedItemId to one element in all tags belonging to one ObjectId using different tag items
+	            ;WITH TagDuplicates (Id, TaggedItemId, first_tagged_item_id, tagged_item_row_num) AS (
+		            SELECT 
+			            t.Id,
+			            t.TaggedItemId,
+			            FIRST_VALUE(t.TaggedItemId) OVER(
+				            PARTITION BY 
+					            [ObjectId], 
+					            [ObjectType]
+				            ORDER BY 
+					            [ObjectId], 
+					            [ObjectType],
+					            t.TaggedItemId
+			            ) as first_tagged_item_id,
+			            ROW_NUMBER() OVER (
+				            PARTITION BY 
+					            [ObjectId], 
+					            [ObjectType]
+				            ORDER BY 
+					            [ObjectId], 
+					            [ObjectType]
+			            ) tagged_item_row_num
+			            FROM [TaggedItem] ti 
+			            INNER JOIN [Tag] t ON ti.Id = t.TaggedItemId
+	            )
+	            UPDATE t
+	            SET 
+		            t.TaggedItemId = TagDuplicates.first_tagged_item_id
+	            FROM Tag t INNER JOIN TagDuplicates ON t.Id = TagDuplicates.Id		
+	            WHERE tagged_item_row_num > 1;
```

Comment:
This might produce redundant updates if one product has multiple tags assigned to the same `TaggedItem`. We might need to add something like `...AND t.TaggedItemId <> first_tagged_item_id` here.

Reply:
Fixed